envoy/config/listener.go

```diff
@@ -50,20 +51,82 @@ func (l *listenerBuilder) addListenerFilterChain(c *listener.FilterChain) *liste
 	return l
 }
 
-func (l *listenerBuilder) AddHTTPManagerFilterChain(httpConnectionManager *hcm.HttpConnectionManager) error {
+func (l *listenerBuilder) addListenerFilter(f *listener.ListenerFilter) *listenerBuilder {
+	l.lr.ListenerFilters = append(l.lr.ListenerFilters, f)
+	return l
+}
+
+type Certificate struct {
+	Cert string
+	Key  string
+}
+
+// AddHTTPManagerFilterChains inserts HTTP Manager as the listener filter chain(s)
+// If certificates are present an additional TLS-enabled filter chain is added and protocol type detection is enabled with TLS Inspector Listener filter.
+func (l *listenerBuilder) AddHTTPManagerFilterChains(httpConnectionManager *hcm.HttpConnectionManager, certs []Certificate) error {
 	anyHTTPManagerConfig, err := anypb.New(httpConnectionManager)
 	if err != nil {
 		return fmt.Errorf("failed to add http manager to the filter chain: cannot convert to Any message type: %w", err)
 	}
-	hcmchain := &listener.FilterChain{
-		Filters: []*listener.Filter{
-			{
-				Name:       wellknown.HTTPConnectionManager,
-				ConfigType: &listener.Filter_TypedConfig{TypedConfig: anyHTTPManagerConfig},
+	hcmFilter := &listener.Filter{
+		Name:       wellknown.HTTPConnectionManager,
+		ConfigType: &listener.Filter_TypedConfig{TypedConfig: anyHTTPManagerConfig},
+	}
+	// Plain HTTP manager filter chain
+	hcmPlainChain := &listener.FilterChain{
+		Filters: []*listener.Filter{hcmFilter},
+	}
+	l.addListenerFilterChain(hcmPlainChain)
+
+	if len(certs) > 0 {
+		// When certificates are present, we add an additional Listener filter chain that is selected when the connection protocol type is tls.
+		// HTTP Manager configuration is the same.
+		// Enable TLS Inspector in the Listener to detect plain http or tls requests.
+		l.addListenerFilter(&listener.ListenerFilter{Name: wellknown.TLSInspector})
+
+		// Make sure plain http manager filter chain is selected when protocol type is raw_buffer (not tls).
+		hcmPlainChain.FilterChainMatch = &listener.FilterChainMatch{TransportProtocol: "raw_buffer"}
```

Comment:
Does this make sense to have ONLY when there are certs? Perhaps we could move it out of the if statement?

Reply:
It is better not to enable something when there is no need for this. So adding TLSInspector and matching on derived raw_buffer needs to be done only when we need to configure TLS